platform/inprocess/mod.rs

```diff
@@ -0,0 +1,359 @@
+        }
+    }
+
+    pub fn try_recv(&self) -> Result<(Vec<u8>, Vec<OpaqueMpscChannel>, Vec<MpscSharedMemory>),MpscError> {
+        let r = self.receiver.borrow();
+        match r.as_ref().unwrap().try_recv() {
+            Ok(MpscChannelMessage(d,c,s)) => Ok((d,
+                                                 c.into_iter().map(|x| OpaqueMpscChannel::new(x)).collect(),
+                                                 s)),
+            Err(_) => Err(MpscError::ChannelClosedError),
+        }
+    }
+}
+
+unsafe impl Send for MpscReceiver { }
+unsafe impl Sync for MpscReceiver { }
+
+#[derive(Clone)]
+pub struct MpscSender {
+    sender: RefCell<mpsc::Sender<MpscChannelMessage>>,
+}
+
+unsafe impl Send for MpscSender { }
+unsafe impl Sync for MpscSender { }
+
+impl MpscSender {
+    fn new(sender: mpsc::Sender<MpscChannelMessage>) -> MpscSender {
+        MpscSender {
+            sender: RefCell::new(sender),
+        }
+    }
+
+    pub fn connect(name: String) -> Result<MpscSender,MpscError> {
+        let record = ONE_SHOT_SENDERS.lock().unwrap().remove(&name).unwrap();
+        record.signal_acquisition();
+        Ok(record.sender)
+    }
+
+    pub fn send(&self,
+                data: &[u8],
+                ports: Vec<MpscChannel>,
+                shared_memory_regions: Vec<MpscSharedMemory>)
+                -> Result<(),MpscError>
+    {
+        match self.sender.borrow().send(MpscChannelMessage(data.to_vec(), ports, shared_memory_regions)) {
+            Err(_) => { Err(MpscError::ChannelClosedError) },
+            Ok(_) => { Ok(()) },
+        }
+    }
+}
+
+pub struct MpscReceiverSet {
+    last_index: usize,
+    receiver_ids: Vec<usize>,
+    receivers: Vec<MpscReceiver>,
+}
+
+impl MpscReceiverSet {
+    pub fn new() -> Result<MpscReceiverSet,MpscError> {
+        Ok(MpscReceiverSet {
+            last_index: 0,
+            receiver_ids: vec![],
+            receivers: vec![],
+        })
+    }
+
+    pub fn add(&mut self, receiver: MpscReceiver) -> Result<i64,MpscError> {
+        self.last_index = self.last_index + 1;
+        self.receiver_ids.push(self.last_index);
+        self.receivers.push(receiver.consume());
+        Ok(self.last_index as i64)
+    }
+
+    pub fn select(&mut self) -> Result<Vec<MpscSelectionResult>,MpscError> {
+        let mut receivers: Vec<Option<mpsc::Receiver<MpscChannelMessage>>> = Vec::with_capacity(self.receivers.len());
+        let mut r_id: i64 = -1;
+        let mut r_index: usize = 0;
+
+        {
+            let select = mpsc::Select::new();
+            // we *must* allocate exact capacity for this, because the Handles *can't move*
+            let mut handles: Vec<mpsc::Handle<MpscChannelMessage>> = Vec::with_capacity(self.receivers.len());
+        
+            for r in &self.receivers {
+                let inner_r = mem::replace(&mut *r.receiver.borrow_mut(), None);
+                receivers.push(inner_r);
+            }
```

Comment:
Instead of swapping out each receiver individually, is it possible to do something like `let receivers = mem::replace(&mut self.receivers, Vec::new()).into_iter().map(Some).collect()`?